### PR TITLE
Medical - Add treatment progress bars and item consumption #VGM-526

### PR DESCRIPTION
**Pull request by veteran29**
- title

~~I'm not happy that the treatment functions are basically duplicated but I do not want to spend time on refactoring it now.~~
Moved into common function.

Whole treatment time/requirement etc. should most likely be described via config or hashmap with all the data.

### CODE REVIEW

File: game/functions/systems/medical/client/fn_medical_itemApplyFAK.sqf [hunk 1]
```diff
@@ -24,5 +25,31 @@ params ["_healer", "_patient", "_bodyPart"];
...
-// TODO progressbar
-["vgm_medical_heal", [_healer, _patient, "fak", _bodyPart], [_patient]] call para_g_fnc_event_triggerTargets;
+private _treatmentTime = 3; // TODO compute the value
+
+[format [localize "STR_VGM_MEDICAL_UI_APPLY_FAK", name _patient], _treatmentTime, {
+    params ["_args", "_startedAt", "_duration"];
+    _args params ["_healer", "_patient"];
+
+    if (_healer distance _patient > 5) exitWith {
+        hint localize "STR_VGM_MEDICAL_UI_NOTIFICATION_PATIENT_TOO_FAR"; // TODO custom notification system?
```
Comment (Spoffy) [inline game/functions/systems/medical/client/fn_medical_itemApplyFAK.sqf]: We can use the custom notification system from MF? It's barely used, and supports a bunch of different things (icons, colours, queued notifications with priorities, etc), I think.

File: game/functions/systems/medical/client/fn_medical_itemApplyFAK.sqf [hunk 2]
```diff
@@ -24,5 +25,31 @@ params ["_healer", "_patient", "_bodyPart"];
...
+        false // return
+    };
+
+    private _requiredItems = vgm_medical_healItems get HEAL_FAK;
+    if (items _healer arrayIntersect _requiredItems isEqualTo []) exitWith {
```
Comment (Spoffy) [inline game/functions/systems/medical/client/fn_medical_itemApplyFAK.sqf]: Might be worth performance testing this if it's going to run often, I recall arrayIntersect being really slow. Not sure how it fares with a 2 item array though, might be fine.

File: game/functions/systems/medical/client/fn_medical_itemApplyMedikit.sqf [hunk 3]
```diff
@@ -24,5 +25,31 @@ params ["_healer", "_patient", "_bodyPart"];
...
+    };
+
+    private _requiredItems = vgm_medical_healItems get HEAL_MEDIKIT;
+    if (items _healer arrayIntersect _requiredItems isEqualTo []) exitWith {
+        hint localize "STR_VGM_MEDICAL_UI_NOTIFICATION_MISSING_ITEMS"; // TODO custom notification system?
+        false // return
+    };
+
+    true // return
+}, {
```
Comment (Spoffy) [inline game/functions/systems/medical/client/fn_medical_itemApplyMedikit.sqf]: Might be worth merging these checks into a function?

**Review by Spoffy — APPROVED** [review]

File: game/functions/systems/medical/client/fn_medical_itemApplyFAK.sqf (hunk 2 above)
Comment (veteran29) [reply game/functions/systems/medical/client/fn_medical_itemApplyFAK.sqf]: Array intersect is actually one of the faster options, but this made me remember about `findAny` which will be the fastest for this use case.

```sqf
// 11 "items" in player inventory
// Execution Time: 0.0014 ms  |  Cycles: 10000/10000  |  Total Time: 14 ms
_requiredItems = ["vn_o_item_firstaidkit", "vn_b_item_firstaidkit"];

items player findAny _requiredItems == -1
```

```sqf
// 11 "items" in player inventory
// Execution Time: 0.0016 ms  |  Cycles: 10000/10000  |  Total Time: 16 ms
_requiredItems = ["vn_o_item_firstaidkit", "vn_b_item_firstaidkit"];

_requiredItems findAny items player == -1
```

```sqf
// 11 "items" in player inventory
// Execution Time: 0.0017 ms  |  Cycles: 10000/10000  |  Total Time: 17 ms
_requiredItems = ["vn_o_item_firstaidkit", "vn_b_item_firstaidkit"];

items player arrayIntersect _requiredItems isEqualTo []
```

```sqf
// 11 "items" in player inventory
// Execution Time: 0.0019 ms  |  Cycles: 10000/10000  |  Total Time: 19 ms
_requiredItems = ["vn_o_item_firstaidkit", "vn_b_item_firstaidkit"];


items player findIf {_x in _requiredItems} == -1
```

```sqf
// 11 "items" in player inventory
// Execution Time: 0.0025 ms  |  Cycles: 10000/10000  |  Total Time: 25 ms
_requiredItems = ["vn_o_item_firstaidkit", "vn_b_item_firstaidkit"];

_items = items player;
_requiredItems findIf {_x in _items} == -1
```

File: game/functions/systems/medical/client/fn_medical_itemApplyFAK.sqf (hunk 1 above)
Comment (veteran29) [reply game/functions/systems/medical/client/fn_medical_itemApplyFAK.sqf]: Maybe later, I do not want to spend the time on this right now.

File: game/functions/systems/medical/client/fn_medical_itemApplyMedikit.sqf (hunk 3 above)
Comment (veteran29) [reply game/functions/systems/medical/client/fn_medical_itemApplyMedikit.sqf]: Refactored item application into a common function and used `findAny` for better perf.

File: game/functions/systems/medical/client/fn_medical_itemApplyFAK.sqf (hunk 1 above)
Comment (Spoffy) [reply game/functions/systems/medical/client/fn_medical_itemApplyFAK.sqf]: Yeah 100%, wasn't thinking now, I should have been clear 🙂

File: game/functions/systems/medical/client/fn_medical_itemApplyFAK.sqf (hunk 2 above)
Comment (Spoffy) [reply game/functions/systems/medical/client/fn_medical_itemApplyFAK.sqf]: Huh, yeah, arrayIntersect did much better than I expected. I think I was throwing around thousands of items in arrays though, so a bit of a different use case 😅 

Thanks for the detailed performance checks though, really appreciate the thoroughness 🙂

File: game/functions/systems/medical/client/fn_medical_itemApplyMedikit.sqf (hunk 3 above)
Comment (Spoffy) [reply game/functions/systems/medical/client/fn_medical_itemApplyMedikit.sqf]: Just checked - perfect, thank you!

**Pull request merged**